### PR TITLE
Updates post text processing to fetch nonexisting custom groups

### PR DESCRIPTION
**Pull request by mgdelacroix**
#### Summary
Updates the logic that gets profiles and statuses for posts to include custom groups fetching as well.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-52964

#### Release Note
```release-note
Fixes bug that wouldn't mark a custom group as a mention if it was not part of the webapp's local state
```

### CODE REVIEW

File: webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts [hunk 1]
```diff
@@ -1112,10 +1115,19 @@ export function getProfilesAndStatusesForPosts(postsArrayOrMap: Post[]|PostList[
...
-        promises.push(getProfilesByUsernames(Array.from(usernamesToLoad))(dispatch, getState));
+    if (usernamesAndGroupsToLoad.size > 0) {
+        // We need to load the profiles synchronously to filter them
+        // out of the groups to check
+        const getProfilesPromise = getProfilesByUsernames(Array.from(usernamesAndGroupsToLoad))(dispatch, getState);
+        promises.push(getProfilesPromise);
+
+        const {data} = await getProfilesPromise as ActionResult<UserProfile[]>;
+        const loadedProfiles = (data || []).map((p) => p.username);
+        const groupsToCheck = Array.from(usernamesAndGroupsToLoad).filter((name) => !loadedProfiles.includes(name));
```
Comment (larkox) [inline webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts:1128]: O(n2) (a complete run over the array, and for each element, a complete run on loadedProfiles)

Using a set should get this to O(n).

File: webapp/channels/src/packages/mattermost-redux/src/actions/groups.ts [hunk 2]
```diff
@@ -156,14 +156,15 @@ export function getGroup(id: string, includeMemberCount = false): ActionFunc {
     });
 }
 
-export function getGroups(filterAllowReference = false, page = 0, perPage = 10, includeMemberCount = false): ActionFunc {
+export function getGroups(q = '', filterAllowReference = false, page = 0, perPage = 10, includeMemberCount = false): ActionFunc {
     return bindClientFunc({
-        clientFunc: async (param1, param2, param3, param4) => {
-            const result = await Client4.getGroups(param1, param2, param3, param4);
+        clientFunc: async (param1, param2, param3, param4, param5) => {
```
Comment (larkox) [inline webapp/channels/src/packages/mattermost-redux/src/actions/groups.ts:161]: Out of the scope, but I don't like this pattern `(param1, param2, ...)`.

File: webapp/platform/client/src/client4.ts [hunk 3]
```diff
@@ -3490,8 +3490,9 @@ export default class Client4 {
         );
     };
 
-    getGroups = (filterAllowReference = false, page = 0, perPage = 10, includeMemberCount = false, hasFilterMember = false) => {
+    getGroups = (q = '', filterAllowReference = false, page = 0, perPage = 10, includeMemberCount = false, hasFilterMember = false) => {
```
Comment (larkox) [inline webapp/platform/client/src/client4.ts:3500]: Why do we need to add this `q` argument? I failed to see where it is used.

**Review by larkox — CHANGES_REQUESTED** [review]

File: webapp/channels/src/packages/mattermost-redux/src/actions/groups.ts (hunk 2 above)
Comment (mgdelacroix) [reply webapp/channels/src/packages/mattermost-redux/src/actions/groups.ts:161]: Agree really

File: webapp/platform/client/src/client4.ts (hunk 3 above)
Comment (mgdelacroix) [reply webapp/platform/client/src/client4.ts:3500]: It's the filter that we set to query groups by name, it's used [here](https://github.com/mattermost/mattermost/pull/24292/files#diff-b0aca3f09658bed4447dad8edfe7dbeb52bed383d70cc15367689883251fafb8R1130)

File: webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts (hunk 1 above)
Comment (mgdelacroix) [reply webapp/channels/src/packages/mattermost-redux/src/actions/posts.ts]: Fixed in https://github.com/mattermost/mattermost/pull/24292/commits/7c3fc6f78b605497096056f3bca06cacd6df2fb9, PTAL

**Review by larkox — APPROVED** [review]

**Review by lindy65 — APPROVED** [review]
@mgdelacroix - thanks, this is fixed, looks good to me 👍

**Review by sbishel — APPROVED** [review]

**Pull request merged**